ntp-daemon/src/system.rs

```diff
@@ -377,137 +365,72 @@ impl<C: NtpClock, T: Wait> System<C, T> {
         }
     }
 
-    fn handle_peer_demobilize(&mut self, index: PeerIndex) {
+    async fn handle_peer_demobilize(&mut self, index: PeerIndex) {
         self.controller.peer_remove(index);
-        self.peers.remove(&index);
+        let state = self.peers.remove(&index).unwrap();
+
+        // Restart the peer reusing its configuration.
+        let spawner_id = state.spawner_id;
+        let peer_id = state.peer_id;
+        let opt_spawner = self.spawners.iter().find(|s| s.id == spawner_id);
+        if let Some(spawner) = opt_spawner {
+            spawner
+                .notify_tx
+                .send(SystemEvent::peer_removed(
+                    peer_id,
+                    PeerRemovalReason::Demobilized,
+                ))
+                .await
+                .expect("Could not notify spawner");
+        }
     }
 
-    fn handle_spawn(
+    async fn create_peer(
         &mut self,
-        peer_address: PeerAddress,
-        addr: SocketAddr,
-        opt_nts: Option<PeerNtsData>,
-    ) {
+        spawner_id: SpawnerId,
+        mut params: PeerCreateParameters,
+    ) -> PeerIndex {
+        info!(peer_id=?params.id, addr=?params.addr, spawner=?spawner_id, "new peer");
         let index = self.peer_indexer.get();
-
         self.peers.insert(
             index,
             PeerState {
                 snapshot: None,
-                peer_address,
+                peer_address: params.normalized_addr.clone(),
+                peer_id: params.id,
+                spawner_id,
             },
         );
         self.controller.peer_add(index);
+
         PeerTask::spawn(
             index,
-            addr,
+            params.addr,
             self.clock.clone(),
             NETWORK_WAIT_PERIOD,
             self.peer_channels.clone(),
-            opt_nts,
+            params.nts.take(),
         );
 
-        // Don't care if there is no receiver
+        // Don't care if there is not receiver
         let _ = self
             .peer_snapshots_sender
             .send(self.observe_peers().collect());
-    }
 
-    #[cfg(test)]
-    fn create_test_peer(&mut self, addr: NormalizedAddress) -> PeerIndex {
-        let index = self.peer_indexer.get();
-
-        self.peers.insert(
-            index,
-            PeerState {
-                snapshot: None,
-                peer_address: PeerAddress::Peer { address: addr },
-            },
-        );
-        self.controller.peer_add(index);
+        // Try and find a related spawner and notify that spawner
+        if let Some(s) = self.spawners.iter().find(|s| s.id == spawner_id) {
+            let _ = s.notify_tx.send(SystemEvent::PeerRegistered(params)).await;
+        }
```

Comment:
This only notifies back to the spawner that the peer is now registered with the system, mostly useful for testing purposes right now. But it can also be used as a feedback mechanism if we ever add a blocking mechanism in system (i.e. we only allow NTS peers in system and block peers that aren't NTS)